src/main/java/org/apache/hadoop/fs/FileSystem.java

```diff
@@ -0,0 +1,4281 @@
+import org.apache.hadoop.io.MultipleIOException;
+import org.apache.hadoop.net.NetUtils;
+import org.apache.hadoop.security.AccessControlException;
+import org.apache.hadoop.security.Credentials;
+import org.apache.hadoop.security.SecurityUtil;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.hadoop.security.token.Token;
+import org.apache.hadoop.security.token.DelegationTokenIssuer;
+import org.apache.hadoop.util.ClassUtil;
+import org.apache.hadoop.util.DataChecksum;
+import org.apache.hadoop.util.Progressable;
+import org.apache.hadoop.util.ReflectionUtils;
+import org.apache.hadoop.util.ShutdownHookManager;
+import org.apache.hadoop.util.StringUtils;
+import org.apache.htrace.core.Tracer;
+import org.apache.htrace.core.TraceScope;
+
+import com.google.common.base.Preconditions;
+import com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static org.apache.hadoop.fs.CommonConfigurationKeysPublic.*;
+
+/****************************************************************
+ * An abstract base class for a fairly generic filesystem.  It
+ * may be implemented as a distributed filesystem, or as a "local"
+ * one that reflects the locally-connected disk.  The local version
+ * exists for small Hadoop instances and for testing.
+ *
+ * <p>
+ *
+ * All user code that may potentially use the Hadoop Distributed
+ * File System should be written to use a FileSystem object or its
+ * successor, {@link FileContext}.
+ *
+ * <p>
+ * The local implementation is {@link LocalFileSystem} and distributed
+ * implementation is DistributedFileSystem. There are other implementations
+ * for object stores and (outside the Apache Hadoop codebase),
+ * third party filesystems.
+ * <p>
+ * Notes
+ * <ol>
+ * <li>The behaviour of the filesystem is
+ * <a href="https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-common/filesystem/filesystem.html">
+ * specified in the Hadoop documentation. </a>
+ * However, the normative specification of the behavior of this class is
+ * actually HDFS: if HDFS does not behave the way these Javadocs or
+ * the specification in the Hadoop documentations define, assume that
+ * the documentation is incorrect.
+ * </li>
+ * <li>The term {@code FileSystem} refers to an instance of this class.</li>
+ * <li>The acronym "FS" is used as an abbreviation of FileSystem.</li>
+ * <li>The term {@code filesystem} refers to the distributed/local filesystem
+ * itself, rather than the class used to interact with it.</li>
+ * <li>The term "file" refers to a file in the remote filesystem,
+ * rather than instances of {@code java.io.File}.</li>
+ * </ol>
+ *****************************************************************/
+@SuppressWarnings("DeprecatedIsStillUsed")
+@InterfaceAudience.Public
+@InterfaceStability.Stable
+public abstract class FileSystem extends Configured
```

Comment:
I'd rather not inline this class, unless we create real running fork of hadoop.

Any alternatives?

Reply:
http://java-performance.info/updating-final-and-static-final-fields/ lists Unsafe as an alternative.

Reply:
The direction of the JVM architects is pretty clear: final fields will no longer be modifiable.  You can see the discussions in these locations:
* https://bugs.openjdk.java.net/browse/JDK-8217225
* https://github.com/powermock/powermock/issues/939
* https://mail.openjdk.java.net/pipermail/core-libs-dev/2018-November/056486.html

If we assume this direction, there are not a lot of choices.  The best would be to get Hadoop patched so we can swap the cache out.  Short of that, we need a patch to the Hadoop code, and I personally feel this is the cleanest technique.  I say that because this library is rarely updated, and when it is updated, it is done (and must be done) by someone that understands the impact of the upgrade.

Reply:
I looked at modifying the bytecode during the shade process, but that has complications from a build process perspective, and also means the source jar will be wrong. This seems like the most straightforward solution.

We already do this with other classes. It means we need to be diligent about applying upstream changes when updating Hadoop versions (which fortunately happens rarely and is a very involved process anyway).

I’ll also see about submitting a patch to Hadoop to make this pluggable.

Reply:
> I’ll also see about submitting a patch to Hadoop to make this pluggable.

very good idea.

> The direction of the JVM architects is pretty clear: final fields will no longer be modifiable. 

i'm aware. But we still have a choice -- accommodate to this as early or as late as possible.

Besides, the advice in https://bugs.openjdk.java.net/browse/JDK-8217225 and link to powermock conflicts with my experience.
I used powermock in the past and I don't want to go there back. 

None of you replies directly to my question, so let me repeat it -- is using Unsafe as way to get away (for the time being)?

Reply:
Based on the links above, it seems that using unsafe to set the static final field will work with current JVMs. However, it's also clear from the comment by Alan Bateman (one of the JDK engineers who worked on modules and modularizing the JDK) that this should not be done:

> The change in JDK 12 is intentional to improve the robustness and security of the platform. In general, there is no supported way to change static final fields. Mocking or other testing tools that want to change static final fields should explore dropping the final field at class load time.

The motivation for this change is that Presto is broken on JDK 12+ due to the existing reflection code. We weren't aware that it was broken until a user reported it. The approach in this PR is future proof. I'd rather not switch from one unsupported hack to another that can break at any point in the future.

Reply:
Another option instead of inlining `FileSystem.java` would be to replace `org.apache.hadoop.fs.FileSystem.Cache` during repackaging, with our own class.

Reply:
I thought about that, but I didn’t think it works because it’s a nested class, so the outer `FileSystem` class would use ours during compilation and tests (none of the needed methods or behavior would exist).

It should work if we could give it the correct name directly, without overriding the outer class. Maybe it can be named with a `$` in the source file name.

Reply:
> so the outer FileSystem class would use ours during compilation and tests (none of the needed methods or behavior would exist).

For our compilation -- it doesn't matter, because `FileSystem` is already compiled.
For our tests -- it's even better, because we test what we ship. (Your impl'd approach also does this, good.)

> Maybe it can be named with a `$` in the source file name.

Yes, this is what i meant. 
Java used to have no problem with that.